subsys/net/ip/net_context.c

```diff
@@ -1187,10 +1187,18 @@ NET_CONN_CB(tcp_established)
 	}
 
 	set_appdata_values(pkt, IPPROTO_TCP);
-	context->tcp->send_ack += net_pkt_appdatalen(pkt);
+
+	uint16_t data_len = net_pkt_appdatalen(pkt);
+	if (data_len > net_tcp_get_recv_wnd(context->tcp)) {
+		NET_ERR("Context %p: overflow of recv window (%d vs %d), pkt dropped",
+			context, net_tcp_get_recv_wnd(context->tcp), data_len);
+		return NET_DROP;
+	}
 
 	ret = packet_received(conn, pkt, context->tcp->recv_user_data);
 
+	context->tcp->send_ack += data_len;
```

Comment:
This is generally more correct. ACK sequence should be updated after we (== app) have received the data, and we (== app) received it after return from the receive callback, not before it.

I'll address the other comments.